### ISSUE TITLE
luajit 2.1 data segment size issues on FreeBSD 11

### ISSUE THREAD

**Issue by maemigh** [LuaJIT/LuaJIT#332]
Having some issues with luajit 2.1 on FreeBSD:
```
$ ./luarocks build cqueues
Warning: falling back to curl - install luasec to get native HTTPS support
Data segment size exceeds process limit
Abort trap
```
Process limit is set quite high though:
```
$ ulimit -a
cpu time               (seconds, -t)  unlimited
file size           (512-blocks, -f)  unlimited
data seg size           (kbytes, -d)  33554432
stack size              (kbytes, -s)  524288
core file size      (512-blocks, -c)  unlimited
max memory size         (kbytes, -m)  unlimited
locked memory           (kbytes, -l)  64
max user processes              (-u)  8499
open files                      (-n)  116937
virtual mem size        (kbytes, -v)  unlimited
swap limit              (kbytes, -w)  unlimited
socket buffer size       (bytes, -b)  unlimited
pseudo-terminals                (-p)  unlimited
kqueues                         (-k)  unlimited
umtx shared locks               (-o)  unlimited
```

The only way I have been able to get around this issue is by setting XCFLAGS+= -DLUAJIT_USE_SYSMALLOC during compile.

**Comment by daurnimator** [comment]
That seems.... odd.
I don't think luarocks would actually be using a huge amount of memory to download/compile/install cqueues.

For the sake of debugging, could you run `luarocks --verbose build cqueues`?

**Comment by maemigh** [comment]
Sure, here it is:

```
./luarocks --verbose build cqueues

io.popen: 	'pwd' 2> /dev/null

os.execute: 	cd '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/bin' && test '-e' '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/lib/luarocks/rocks'
Results: 1
  1 (number): 256

os.execute: 	cd '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/bin' && test '-e' '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit'
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/bin' && test '-d' '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit'
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/bin' && test '-e' '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit'
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/bin' && mkdir -p '/home/maemigh/.cache/luarocks/https___luarocks.org'
Results: 1
  1 (number): 0
Warning: falling back to curl - install luasec to get native HTTPS support

os.execute: 	cd '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/bin' && 'curl' -k -f -L --user-agent "LuaRocks/2.4.2 freebsd-x86_64 via curl" --connect-timeout 30 'https://luarocks.org/manifest-5.1.zip' > '/home/maemigh/.cache/luarocks/https___luarocks.org/manifest-5.1.zip' 2> /dev/null
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/bin' && test '-d' '/home/maemigh/.cache/luarocks/https___luarocks.org'
Results: 1
  1 (number): 0

os.execute: 	cd '/home/maemigh/.cache/luarocks/https___luarocks.org' && rm '-rf' '/home/maemigh/.cache/luarocks/https___luarocks.org/manifest-5.1'
Results: 1
  1 (number): 0

os.execute: 	cd '/home/maemigh/.cache/luarocks/https___luarocks.org' && unzip -n '/home/maemigh/.cache/luarocks/https___luarocks.org/manifest-5.1.zip'
Archive:  /home/maemigh/.cache/luarocks/https___luarocks.org/manifest-5.1.zip
 extracting: manifest-5.1  
Results: 1
  1 (number): 0

io.popen: 	mktemp -d '/tmp/luarocks_luarocks-rock-cqueues-20161215.51-0-XXXXXX'

os.execute: 	cd '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/bin' && test '-d' '/tmp/luarocks_luarocks-rock-cqueues-20161215.51-0-Fbu5Du'
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/luarocks_luarocks-rock-cqueues-20161215.51-0-Fbu5Du' && 'curl' -k -f -L --user-agent "LuaRocks/2.4.2 freebsd-x86_64 via curl" --connect-timeout 30 'https://luarocks.org/cqueues-20161215.51-0.src.rock' > '/tmp/luarocks_luarocks-rock-cqueues-20161215.51-0-Fbu5Du/cqueues-20161215.51-0.src.rock' 2> /dev/null
Results: 1
  1 (number): 0

io.popen: 	mktemp -d '/tmp/luarocks_cqueues-20161215.51-0-XXXXXX'

os.execute: 	cd '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/bin' && test '-d' '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2'
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2' && unzip -n '/tmp/luarocks_luarocks-rock-cqueues-20161215.51-0-Fbu5Du/cqueues-20161215.51-0.src.rock'
Archive:  /tmp/luarocks_luarocks-rock-cqueues-20161215.51-0-Fbu5Du/cqueues-20161215.51-0.src.rock
 extracting: cqueues-20161215.51-0.rockspec  
 extracting: rel-20161215.tar.gz  
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/bin' && test '-d' '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2'
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2' && test '-e' 'cqueues-20161215.51-0.rockspec'
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2' && test '-f' '/usr/local/include/openssl/crypto.h'
Results: 1
  1 (number): 256

os.execute: 	cd '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2' && test '-f' '/usr/local/include/openssl/crypto.h'
Results: 1
  1 (number): 256

os.execute: 	cd '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2' && test '-f' '/usr/include/openssl/crypto.h'
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2' && test '-f' '/usr/lib/libcrypto.a'
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2' && test '-f' '/usr/local/include/openssl/ssl.h'
Results: 1
  1 (number): 256

os.execute: 	cd '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2' && test '-f' '/usr/local/include/openssl/ssl.h'
Results: 1
  1 (number): 256

os.execute: 	cd '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2' && test '-f' '/usr/include/openssl/ssl.h'
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2' && test '-f' '/usr/lib/libssl.a'
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2' && test '-d' '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/lib/luarocks/rocks/cqueues/20161215.51-0'
Results: 1
  1 (number): 256

os.execute: 	cd '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2' && gunzip -c 'rel-20161215.tar.gz' | tar -xf -
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2' && test '-d' 'cqueues-rel-20161215'
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2/cqueues-rel-20161215' && mkdir -p '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/lib/luarocks/rocks/cqueues/20161215.51-0/lua'
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2/cqueues-rel-20161215' && mkdir -p '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/lib/luarocks/rocks/cqueues/20161215.51-0/bin'
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2/cqueues-rel-20161215' && mkdir -p '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/lib/luarocks/rocks/cqueues/20161215.51-0/conf'
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2/cqueues-rel-20161215' && mkdir -p '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/lib/luarocks/rocks/cqueues/20161215.51-0/lib'
Results: 1
  1 (number): 0

os.execute: 	cd '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2/cqueues-rel-20161215' && gmake -f GNUmakefile all5.1 'CC=cc' 'LDFLAGS=-shared -L"/usr/lib" -L"/usr/lib"' 'libdir=/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/lib' 'includedir=/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/include/luajit-2.1' 'bindir=/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/bin' 'CPPFLAGS=-I"/usr/include" -I"/usr/include"' 'CFLAGS=-O2 -fPIC'
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
cp /tmp/luarocks_cqueues-20161215.51-0-5MnkB2/cqueues-rel-20161215/config.h.guess /tmp/luarocks_cqueues-20161215.51-0-5MnkB2/cqueues-rel-20161215/config.h
cp /tmp/luarocks_cqueues-20161215.51-0-5MnkB2/cqueues-rel-20161215/config.h /tmp/luarocks_cqueues-20161215.51-0-5MnkB2/cqueues-rel-20161215/src/config.h
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
Data segment size exceeds process limit
Abort trap
mkdir -p /tmp/luarocks_cqueues-20161215.51-0-5MnkB2/cqueues-rel-20161215/src/5.1
cc -O2 -fPIC -DLUA_COMPAT_APIINTCASTS  -D_REENTRANT -D_THREAD_SAFE -D_GNU_SOURCE -I"/usr/include" -I"/usr/include" -DCQUEUES_VENDOR='"william@25thandClement.com"' -DCQUEUES_VERSION=20161215L  -c -o /tmp/luarocks_cqueues-20161215.51-0-5MnkB2/cqueues-rel-20161215/src/5.1/cqueues.o /tmp/luarocks_cqueues-20161215.51-0-5MnkB2/cqueues-rel-20161215/src/cqueues.c
Data segment size exceeds process limit
gmake: *** [/tmp/luarocks_cqueues-20161215.51-0-5MnkB2/cqueues-rel-20161215/src/GNUmakefile:76: /tmp/luarocks_cqueues-20161215.51-0-5MnkB2/cqueues-rel-20161215/src/5.1/cqueues.o] Abort trap
Results: 1
  1 (number): 512

Error: Build error: Failed building.

os.execute: 	cd '/' && rm '-rf' '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/lib/luarocks/rocks/cqueues/20161215.51-0'
Results: 1
  1 (number): 0

os.execute: 	cd '/' && rmdir '/tmp/output-luajit-freebsd64/pkgbuild/usr/local/luajit/lib/luarocks/rocks/cqueues'
Results: 1
  1 (number): 0

os.execute: 	cd '/' && rm '-rf' '/tmp/luarocks_cqueues-20161215.51-0-5MnkB2'
Results: 1
  1 (number): 0

os.execute: 	cd '/' && rm '-rf' '/tmp/luarocks_luarocks-rock-cqueues-20161215.51-0-Fbu5Du'
Results: 1
  1 (number): 0
```

**Comment by daurnimator** [comment]
So as far as a I can see, luajit adjusts the process rlimit on freebsd https://github.com/LuaJIT/LuaJIT/blob/8ada57eb49fb03d4d1c3cb37e534b97fbb92a5e6/src/lj_alloc.c#L228
That then gets inherited by child processes.

This is annoying... I'm not sure what the fix would be aside from restoring the rlimit after the `mmap` call.
Luckily, it [looks like newer freebsds support the MAP_32BIT flag to mmap](https://github.com/freebsd/freebsd/commit/04bb6e10cd2cbd0ee0de7fef3f81e88ebd43e6b4), and hence for newer releases we don't need to use `setrlimit` any more.

**Comment by maemigh** [comment]
Commenting out those lines seems to have resolved the problem.

**Comment by daurnimator** [comment]
> Commenting out those lines seems to have resolved the problem.

Are you on freebsd >= 11.0? A fix for you could be as simple as adding a version check for >= 11 to the `#if` at https://github.com/LuaJIT/LuaJIT/blob/8ada57eb49fb03d4d1c3cb37e534b97fbb92a5e6/src/lj_alloc.c#L223

**Comment by johnae** [comment]
I've also run into this and submitted a pull request for it here: https://github.com/LuaJIT/LuaJIT/pull/333

**Comment by MikePall** [comment]
Fixed.

**Comment by myfreeweb** [comment]
The fix in 2.1 is wrong.

https://github.com/LuaJIT/LuaJIT/blob/v2.1/src/lj_alloc.c#L347-L357

`((defined(__FreeBSD__) && __FreeBSD__ < 10) || defined(__FreeBSD_kernel__))`

`__FreeBSD_kernel__` is still defined on any version (I'm on -CURRENT right now), so `setrlimit` was still used, and I couldn't run anything heavier than a shell inside neovim. Also, you don't need `__FreeBSD_kernel__`.
